CalibPPS/ESProducers/plugins/CTPPSLHCInfoRandomXangleESSource.cc

```diff
@@ -0,0 +1,145 @@
+
+CTPPSLHCInfoRandomXangleESSource::CTPPSLHCInfoRandomXangleESSource(const edm::ParameterSet &conf)
+    : m_label(conf.getParameter<std::string>("label")),
+
+      m_generateEveryNEvents(conf.getParameter<unsigned int>("generateEveryNEvents")),
+
+      m_beamEnergy(conf.getParameter<double>("beamEnergy")),
+      m_betaStar(conf.getParameter<double>("betaStar")) {
+  const auto &xangleHistogramFile = conf.getParameter<std::string>("xangleHistogramFile");
+  const auto &xangleHistogramObject = conf.getParameter<std::string>("xangleHistogramObject");
+
+  TFile *f_in = TFile::Open(xangleHistogramFile.c_str());
+  TH1D *h_xangle = (TH1D *)f_in->Get(xangleHistogramObject.c_str());
+
+  double s = 0.;
+  for (int bi = 1; bi <= h_xangle->GetNbinsX(); ++bi)
+    s += h_xangle->GetBinContent(bi);
+
+  double cw = 0.;
+  for (int bi = 1; bi <= h_xangle->GetNbinsX(); ++bi) {
+    double xangle = h_xangle->GetBinCenter(bi);
+    double w = h_xangle->GetBinContent(bi) / s;
+
+    binData.push_back({cw, cw + w, xangle});
+
+    cw += w;
+  }
+
+  delete f_in;
+
+  setWhatProduced(this, m_label);
+  findingRecord<LHCInfoRcd>();
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void CTPPSLHCInfoRandomXangleESSource::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<std::string>("label", "")->setComment("label of the LHCInfo record");
+
+  desc.add<unsigned int>("generateEveryNEvents", 1)->setComment("how often to generate new xangle");
+
+  desc.add<std::string>("xangleHistogramFile", "")->setComment("ROOT file with xangle distribution");
+  desc.add<std::string>("xangleHistogramObject", "")->setComment("xangle distribution object in the ROOT file");
+
+  desc.add<double>("beamEnergy", 0.)->setComment("beam energy");
+  desc.add<double>("betaStar", 0.)->setComment("beta*");
+
+  descriptions.add("ctppsLHCInfoRandomXangleESSource", desc);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void CTPPSLHCInfoRandomXangleESSource::setIntervalFor(const edm::eventsetup::EventSetupRecordKey &key,
+                                                      const edm::IOVSyncValue &iosv,
+                                                      edm::ValidityInterval &oValidity) {
+  edm::EventID beginEvent = iosv.eventID();
+  edm::EventID endEvent(beginEvent.run(), beginEvent.luminosityBlock(), beginEvent.event() + m_generateEveryNEvents);
+  oValidity = edm::ValidityInterval(edm::IOVSyncValue(beginEvent), edm::IOVSyncValue(endEvent));
+}
+
+//----------------------------------------------------------------------------------------------------
+
+edm::ESProducts<std::unique_ptr<LHCInfo>> CTPPSLHCInfoRandomXangleESSource::produce(const LHCInfoRcd &) {
+  auto output = std::make_unique<LHCInfo>();
+
+  // TODO: this doesn't work, use temporarily ROOT code
```

Comment:
@jan-kaspar @pohsun why all this? There is no reproducibility guaranteed in this code as far as I can see...

Reply:
Sorry, I should have pre-discussed with you. As you may know, the xangle has an important impact on what/how PPS can see. I've been asked by the POG to make a simulation where the xangle would be generated randomly according to a specified histogram (e.g. as from LHC data for a given JSON selection). My attempt is the code currently in the PR and I acknowledge it has issues. What would be your recommendation?